### PR TITLE
V2 anonymizer dto

### PR DESCRIPTION
**Pull request by shiranr**

### CODE REVIEW

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py [hunk 1]
```diff
@@ -27,19 +25,21 @@ def __init__(
         :param data: a map which contains the anonymizers, analyzer_results and text
         """
 
-    def anonymize(self, engine_request: AnonymizerRequest) -> str:
+    def anonymize(self, text: str, engine_request: AnonymizerRequest) -> str:
```
Comment (navalev) [inline presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:28]:
```suggestion
    def anonymize(self, text: str, anonymize_request: AnonymizerRequest) -> str:
```

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py [hunk 2]
```diff
@@ -48,31 +48,55 @@ def anonymize(self, engine_request: AnonymizerRequest) -> str:
...
+        params = anonymizer_config.params
+        params["entity_type"] = entity_type
         anonymized_text = anonymizer.anonymize(
-            params=anonymizer_dto, text=text_to_anonymize
+            params=params, text=text_to_anonymize
         )
         return anonymized_text
+
+    @staticmethod
+    def get_anonymizer(anonymizers: dict, entity_type: str):
```
Comment (navalev) [inline presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:87]: shouldnt this method be private?
```suggestion
    def get_anonymizer_by_entity_type(anonymizers: dict, entity_type: str):
```

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py [hunk 3]
```diff
@@ -48,31 +48,55 @@ def anonymize(self, engine_request: AnonymizerRequest) -> str:
...
+        :param anonymizers: a dictionary - key - entity_type we are replacing
+        value - AnonymizerConfig
+        :param entity_type: the type of the text we want to do anonymizer over
+        :return: AnonymizerConfig class
+        """
+        anonymizer = anonymizers.get(entity_type)
+        if not anonymizer:
+            anonymizer = anonymizers.get("DEFAULT")
+            if not anonymizer:
+                anonymizer = AnonymizerConfig("replace", {})
```
Comment (navalev) [inline presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:101]: "replace" better by a const like string, in case we want to replace the default

**Review by navalev — CHANGES_REQUESTED** [review]
so u're not using a seperate class for dto. LGTM, with some comments mostly around naming. We still need to update the samples with this change but will be done in a seperate PR

File: presidio-anonymizer/presidio_anonymizer/entities/anonymizer_request.py [hunk 4]
```diff
@@ -8,51 +8,31 @@
...
 
     def get_analysis_results(self):
         """Get the analysis results."""
         return self._analysis_results
 
+    def get_anonymizers_config(self):
+        """Get the anonymizers data transfer objects."""
+        return self._anonymizers_config
+
     def __validate_and_insert_input(self, data: dict):
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/anonymizer_request.py:36]: In the next PR this is gone :) but I entered it...

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py (hunk 1 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:28]: This is going to be removed next PR.

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py (hunk 3 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py]: Done.

File: presidio-anonymizer/presidio_anonymizer/entities/anonymizer_config.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            "from_end": true
+        }
+        :return: AnonymizerConfig
+        """
+        anonymizer_name = params.get("type")
+        params.pop("type")
+        return cls(anonymizer_name, params)
+
+    def __init__(self, anonymizer_name: str, params: dict = {}):
+        self.logger = logging.getLogger("presidio-anonymizer")
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/anonymizer_config.py:29]: Where I could I added it to the init, if the class doesn't have an init, I left it in the class level

**Review by navalev — APPROVED** [review]

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py [hunk 6]
```diff
@@ -14,32 +17,31 @@ class AnonymizerEngine:
...
-        cls.anonymizer_name(cls): cls for cls in Anonymizer.__subclasses__()
-    }
-
     def __init__(
             self,
+
     ):
         """Handle text replacement for PIIs with requested anonymizers.
 
         :param data: a map which contains the anonymizers, analyzer_results and text
```
Comment (TessFerrandez) [inline presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:26]:
```suggestion
```

Comment (TessFerrandez) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:26]: the parameter no longer exists

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py [hunk 7]
```diff
@@ -48,31 +50,55 @@ def anonymize(self, engine_request: AnonymizerRequest) -> str:
...
-                                                                      text_to_anonymize)
+            anonymized_text = self.__extract_anonymizer_and_anonymize(
+                analyzer_result.entity_type, anonymizer_config,
+                text_to_anonymize)
             text_builder.replace_text(anonymized_text, analyzer_result.start,
                                       analyzer_result.end)
 
         return text_builder.output_text
 
     def anonymizers(self):
```
Comment (TessFerrandez) [inline presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:70]:
```suggestion
    @staticmethod
    def anonymizers():
```

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py [hunk 8]
```diff
@@ -48,31 +50,55 @@ def anonymize(self, engine_request: AnonymizerRequest) -> str:
...
         """Return a list of supported anonymizers."""
-        names = [p for p in self.builtin_anonymizers.keys()]
+        names = [p for p in Anonymizer.get_anonymizers().keys()]
         return names
 
-    def __extract_anonymizer_and_anonymize(self, anonymizer_dto, text_to_anonymize):
-        anonymizer = anonymizer_dto.get("anonymizer")()
+    def __extract_anonymizer_and_anonymize(self, entity_type: str,
```
Comment (TessFerrandez) [inline presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:75]:
```suggestion
    @staticmethod
    def __extract_anonymizer_and_anonymize(entity_type: str,
```

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py (hunk 7 above)
Comment (TessFerrandez) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:70]: Also... should this be called get_anonymizers() for consistency?

File: presidio-anonymizer/presidio_anonymizer/anonymizers/anonymizer.py [hunk 9]
```diff
@@ -22,3 +24,13 @@ def validate(self, params: dict = None) -> None:
     def anonymizer_name(self) -> str:
         """Return anonymizer name."""
         pass
+
+    @staticmethod
+    def get_anonymizers():
+        """Return all anonymizers classes currently available."""
+        if not Anonymizer._anonymizers:
+            Anonymizer._anonymizers = {
+                cls.anonymizer_name(cls): cls
```
Comment (TessFerrandez) [inline presidio-anonymizer/presidio_anonymizer/anonymizers/anonymizer.py:33]: In pycharm, this is complaining that it "Expected type 'Anonymizer' but got 'Type[Anonymizer]' instead" which seems to make sense... is this working as it should?

File: presidio-anonymizer/presidio_anonymizer/entities/anonymizer_request.py [hunk 10]
```diff
@@ -87,31 +63,14 @@ def __handle_anonymizers(self, data):
...
-                anonymizer_dto["anonymizer"] = anonymizer
-                self._anonymizers[key] = anonymizer_dto
+            for key, anonymizer_json in anonymizers.items():
+                self.logger.debug(
+                    f"converting {anonymizer_json} to anonymizer config class")
+                anonymizer_config = AnonymizerConfig.from_json(anonymizer_json)
+                self._anonymizers_config[key] = anonymizer_config
 
     def __handle_text(self, data):
         self._text = data.get("text")
```
Comment (TessFerrandez) [inline presidio-anonymizer/presidio_anonymizer/entities/anonymizer_request.py:73]: all other functions have type hints, and doc strings, should this also have it

File: presidio-anonymizer/presidio_anonymizer/entities/anonymizer_request.py [hunk 11]
```diff
@@ -8,55 +8,31 @@
...
 
-    def __validate_and_insert_input(self, data: dict):
-        self.__handle_text(data)
-        self.__handle_analyzer_results(data)
-        self.__handle_anonymizers(data)
+    def get_anonymizers_config(self):
+        """Get the anonymizers data transfer objects."""
+        return self._anonymizers_config
 
     def __handle_analyzer_results(self, data):
```
Comment (TessFerrandez) [inline presidio-anonymizer/presidio_anonymizer/entities/anonymizer_request.py:37]: all other functions have type hints, should we add it here too

File: presidio-anonymizer/tests/anonymizers/test_anonymizer.py [hunk 12]
```diff
@@ -0,0 +1,8 @@
+from presidio_anonymizer.anonymizers import Anonymizer
+
+
+def test_given_anonymizers_list_then_all_classes_are_there():
+    anonymizers = Anonymizer.get_anonymizers()
+    assert len(anonymizers) >= 5
+    for class_name in ["hash", "mask", "redact", "replace", "fpe"]:
+        assert anonymizers.get(class_name)
```
Comment (TessFerrandez) [inline presidio-anonymizer/tests/anonymizers/test_anonymizer.py:8]: what is this asserting? does get return None if it doesn't exist?

**Review by TessFerrandez — APPROVED** [review]
looks good, mostly nits and some functions that could be made static

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py (hunk 7 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:70]: Will be done in next PR (535)

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py (hunk 8 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:75]: I am using the logger here... I can't do that.

File: presidio-anonymizer/presidio_anonymizer/entities/anonymizer_request.py (hunk 10 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/anonymizer_request.py:73]: return type yes, doctsring no (only on public methods).

Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/anonymizer_request.py:73]: but this method does not exist anymore :) in next PR

File: presidio-anonymizer/tests/anonymizers/test_anonymizer.py (hunk 12 above)
Comment (shiranr) [reply presidio-anonymizer/tests/anonymizers/test_anonymizer.py:8]: Done in next PR

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py (hunk 6 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:26]: Will be done in next PR (535)

**Comment by shiranr** [comment]
This was pushed with 535.

**Pull request closed without merging**